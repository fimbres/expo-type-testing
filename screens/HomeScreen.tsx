import React, { FC } from 'react'
import { StyleSheet, Text, View } from 'react-native'
import { NativeStackScreenProps } from '@react-navigation/native-stack'

import { Styles } from '../constants'
import { RootStackParamList } from '../navigation/types'

const HomeScreen: FC<NativeStackScreenProps<RootStackParamList, "Home">> = () => {
  return (
    <View style={styles.container}>
      <Text style={Styles.textTitleGiant}>HomeScreen</Text>
      <Text style={Styles.textCaptionExtraSmall}>HomeScreen</Text>
    </View>
  )
}

export default HomeScreen;

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center'
    }
});
