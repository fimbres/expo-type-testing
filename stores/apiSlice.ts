import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

import { IProduct } from './types';

export const apiSlice = createApi({
    reducerPath: 'api',
    tagTypes: ['Product'],
    baseQuery: fetchBaseQuery({
        baseUrl: 'https://6222994f666291106a29f999.mockapi.io/api/v1'
    }),
    endpoints: builder => ({
        getProducts: builder.query({
            query: () => '/products',
            transformResponse: loadedProducts => {
                return loadedProducts as IProduct[];
            }
        })
    })
});

export const { useGetProductsQuery } = apiSlice;
