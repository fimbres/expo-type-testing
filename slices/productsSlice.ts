import { createSelector } from "@reduxjs/toolkit";
import moment from "moment";

import { apiSlice } from "./apiSlice";
import { IProduct } from "../types";

const initialState: IProduct[] = [];

const selectUsersResult = apiSlice.endpoints.getProducts.select({});

const sortProducts = (first: IProduct, second: IProduct) => {
  return moment(second.createdAt).valueOf() - moment(first.createdAt).valueOf();
};

export const selectAllProducts = createSelector(
  selectUsersResult,
  (productsResult) =>
    productsResult.data
      ? [...productsResult.data].sort(sortProducts)
      : initialState
);

export const selectTotalPoints = createSelector(selectAllProducts, (products) =>
  products.reduce((sum, product) => {
    if (product.is_redemption) {
      return (sum -= product.points);
    } else {
      return (sum += product.points);
    }
  }, 0)
);

export const selectProductById = createSelector(
  selectAllProducts,
  (_: IProduct[], productId: number) => productId,
  (products, productId) => products.find((product) => product.id == productId)
);

export const selectRedemptionProducts = createSelector(
  selectAllProducts,
  (products) => products.filter((product) => product.is_redemption)
);

export const selectNotRedemptionProducts = createSelector(
  selectAllProducts,
  (products) => products.filter((product) => !product.is_redemption)
);
