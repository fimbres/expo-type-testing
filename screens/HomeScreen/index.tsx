import React, { FC, useEffect, useState } from "react";
import { StyleSheet, Text, View } from "react-native";
import { NativeStackScreenProps } from "@react-navigation/native-stack";
import { useSelector } from "react-redux";
import Toast from "react-native-root-toast";

import { Colors, Styles } from "../../constants";
import {
  FilterButtons,
  Loader,
  PointsBadge,
  ProductsList,
  SafeAreaContainer,
  ScreenHeader,
  ScreenFooter,
} from "../../components";
import { Filter } from "../../types";
import { RootStackParamList } from "../../navigation/types";
import { useGetProductsQuery } from "../../slices/apiSlice";
import {
  selectTotalPoints,
  selectAllProducts,
  selectNotRedemptionProducts,
  selectRedemptionProducts,
} from "../../slices/productsSlice";

const HomeScreen: FC<NativeStackScreenProps<RootStackParamList, "Home">> = ({
  navigation,
  route,
}) => {
  const { isError, isLoading, error } = useGetProductsQuery({});
  const totalPoints = useSelector(selectTotalPoints);
  const allProducts = useSelector(selectAllProducts);
  const redemptionProducts = useSelector(selectRedemptionProducts);
  const notRedemptionProducts = useSelector(selectNotRedemptionProducts);
  const [filter, setFilter] = useState<Filter>(
    route.params?.initialFilter || "all"
  );

  useEffect(() => {
    if (isError) {
      showToast();
      console.error("Error: ", error);
    }
  }, [isError]);

  const showToast = () => {
    Toast.show("Error cargando los productos", {
      duration: Toast.durations.LONG,
      containerStyle: {
        marginTop: 40,
        backgroundColor: Colors.red,
      },
      textColor: Colors.white,
      position: 1,
    });
  };

  const getProducts = () => {
    switch (filter) {
      case "all":
        return allProducts;
      case "redemption":
        return redemptionProducts;
      case "no-redemption":
        return notRedemptionProducts;
    }
  };

  return (
    <SafeAreaContainer>
      <>
        {isLoading && <Loader />}
        <ScreenHeader
          title="Bienvenido de vuelta!"
          subtitle="Ruben Rodriguez"
        />
        <View style={styles.container}>
          <Text style={styles.subTitle}>Tus Puntos</Text>
          <PointsBadge points={totalPoints} />
          <Text style={styles.subTitle}>Tus Movimientos</Text>
          <ProductsList
            products={getProducts()}
            onPress={(product) => navigation.navigate("Product", { product })}
          />
          <ScreenFooter>
            <FilterButtons filter={filter} setFilter={setFilter} />
          </ScreenFooter>
        </View>
      </>
    </SafeAreaContainer>
  );
};

export default HomeScreen;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingHorizontal: 20,
  },
  subTitle: {
    ...Styles.textTitleSmall,
    color: Colors.gray,
    textTransform: "uppercase",
    marginVertical: 20,
  },
});
