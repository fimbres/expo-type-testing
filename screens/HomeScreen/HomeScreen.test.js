import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent, act } from '@testing-library/react-native';
import { ApiProvider } from '@reduxjs/toolkit/dist/query/react';

import HomeScreen from '.';
import { apiSlice } from '../../slices/apiSlice';

describe('<HomeScreen />', () => {
    it('render text correctly', () => {
        render(
            <ApiProvider api={apiSlice}>
                <HomeScreen route={{ params: { initialFilter: "all" }}} />
            </ApiProvider>
        );

        const firstText = screen.getByText('Bienvenido de vuelta!');
        const secondText = screen.getByText('Ruben Rodriguez');
        const thirdText = screen.getByText('Tus Puntos');
        const fourthText = screen.getByText('Tus Movimientos');

        expect(firstText).toBeDefined();
        expect(secondText).toBeDefined();
        expect(thirdText).toBeDefined();
        expect(fourthText).toBeDefined();
    });

    it('press filter buttons', () => {
        const { getAllByTestId } = render(
            <ApiProvider api={apiSlice}>
                <HomeScreen route={{ params: { initialFilter: "redemption" }}}/>
            </ApiProvider>
        );

        const button = getAllByTestId('app-button');

        act(() => {
            fireEvent.press(button[0]);
        })
    });

    it('renders correctly', () => {
        const tree = renderer
        .create(
            <ApiProvider api={apiSlice}>
                <HomeScreen route={{ params: { initialFilter: "all" }}} />
            </ApiProvider>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
