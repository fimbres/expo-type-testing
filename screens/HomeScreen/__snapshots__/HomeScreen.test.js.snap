// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeScreen /> renders correctly 1`] = `
<RCTSafeAreaView
  style={
    Array [
      Object {
        "backgroundColor": "#F8F8F8",
        "flex": 1,
        "paddingTop": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#020202CC",
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 5,
      }
    }
    testID="loader"
  >
    <ActivityIndicator
      color="#CFD6FF"
      size="large"
    />
  </View>
  <View
    style={
      Array [
        Object {
          "paddingHorizontal": 20,
          "paddingTop": 24,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#020202",
          "fontFamily": "avenir-black",
          "fontSize": 20,
          "fontWeight": "800",
          "lineHeight": 27.32,
        }
      }
    >
      Bienvenido de vuelta!
    </Text>
    <Text
      style={
        Object {
          "color": "#020202",
          "fontFamily": "avenir-book",
          "fontSize": 16,
          "fontWeight": "400",
          "lineHeight": 21.86,
        }
      }
    >
      Ruben Rodriguez
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "paddingHorizontal": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#9B9898",
          "fontFamily": "avenir-black",
          "fontSize": 14,
          "fontWeight": "800",
          "lineHeight": 19.12,
          "marginVertical": 20,
          "textTransform": "uppercase",
        }
      }
    >
      Tus Puntos
    </Text>
    <View
      style={
        Object {
          "backgroundColor": "#334FFA",
          "borderRadius": 20,
          "elevation": 3,
          "marginHorizontal": 32,
          "paddingBottom": 49,
          "paddingHorizontal": 18,
          "paddingTop": 21,
          "shadowColor": "#020202",
          "shadowOffset": Object {
            "height": 3,
            "width": 2,
          },
          "shadowOpacity": 0.5,
          "shadowRadius": 8,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontFamily": "avenir-black",
            "fontSize": 16,
            "fontWeight": "800",
            "lineHeight": 21.86,
            "marginBottom": 7,
          }
        }
      >
        Diciembre
      </Text>
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontFamily": "avenir-black",
            "fontSize": 32,
            "fontWeight": "800",
            "lineHeight": 43.71,
            "textAlign": "center",
          }
        }
      >
        0.00
         pts
      </Text>
    </View>
    <Text
      style={
        Object {
          "color": "#9B9898",
          "fontFamily": "avenir-black",
          "fontSize": 14,
          "fontWeight": "800",
          "lineHeight": 19.12,
          "marginVertical": 20,
          "textTransform": "uppercase",
        }
      }
    >
      Tus Movimientos
    </Text>
    <RCTScrollView
      ItemSeparatorComponent={[Function]}
      ListEmptyComponent={[Function]}
      ListFooterComponent={[Function]}
      data={Array []}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 16,
          "flex": 1,
          "paddingVertical": 23,
          "width": "100%",
        }
      }
      testID="products-list"
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "marginTop": 25,
            }
          }
          testID="products-list-empty-state"
        >
          <Image
            source={1}
            style={
              Object {
                "height": 150,
                "marginBottom": 24,
                "resizeMode": "cover",
                "width": 190,
              }
            }
          />
          <Text
            style={
              Object {
                "color": "#020202",
                "fontFamily": "avenir-black",
                "fontSize": 16,
                "fontWeight": "800",
                "lineHeight": 21.86,
              }
            }
          >
            No hay movimientos registrados
          </Text>
          <Text
            style={
              Object {
                "color": "#9B9898",
                "fontFamily": "avenir-book",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 21.86,
              }
            }
          >
            Realiza una compra y vuelve m√°s tarde
          </Text>
        </View>
        <View
          onLayout={[Function]}
        >
          <View
            style={
              Object {
                "marginTop": 33,
              }
            }
          />
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "marginVertical": 40,
          },
          undefined,
        ]
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "#334FFA",
              "borderRadius": 10,
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
            },
            Object {
              "paddingVertical": 17,
            },
            false,
          ]
        }
        testID="app-button"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#020202",
                "fontFamily": "avenir-black",
                "fontSize": 12,
                "fontWeight": "800",
                "lineHeight": 16.39,
              },
              Object {
                "color": "#FFFFFF",
              },
            ]
          }
        >
          Ganados
        </Text>
      </View>
      <View
        style={
          Object {
            "marginHorizontal": 6,
          }
        }
      />
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "#334FFA",
              "borderRadius": 10,
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
            },
            Object {
              "paddingVertical": 17,
            },
            false,
          ]
        }
        testID="app-button"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#020202",
                "fontFamily": "avenir-black",
                "fontSize": 12,
                "fontWeight": "800",
                "lineHeight": 16.39,
              },
              Object {
                "color": "#FFFFFF",
              },
            ]
          }
        >
          Canjeados
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
